Improve inline documentation of config defaults and use it to generate website counterpart

astorije opened the pull request:
This is a first part, pre-requisite for https://github.com/thelounge/thelounge.github.io/pull/84.

So the way it works is that the script generates a Markdown from all the inline comments and puts it in [`_includes/config.js.md` of the doc folder](https://github.com/thelounge/thelounge.github.io/pull/84/files#diff-df923e880eb889dae72e108ad9bb7165). The generated file needs to be checked in and it is then read [by `_docs/configuration.md`](https://github.com/thelounge/thelounge.github.io/pull/84/files#diff-5b88d6756daca0ac2a4e130b4121f431R19).
When the website gets built, it uses the same content than the actual config file to reduce [this kind of totally out-of-sync mess](https://thelounge.chat/docs/server/configuration.html).
It then looks like this: [configuration.html](https://deploy-preview-84--thelounge-chat.netlify.com/docs/configuration.html). Note that my PR is not up-to-date with the latest defaults as we speak. Also I think this page needs needs a TOC but that would affect the website/docs layout and not this file/PR.

I know this isn't perfect, but hey, it's a first step :) We can do a lot of things to improve keeping them in sync, but my goal was to keep the doc both in the config file _and_ the website while trying to keep them in sync more easily than it is at the moment.

I have checked that my rewrite of `defaults/config.js` is 1:1 in actual code with `master`. Keys are in a different order, but all the keys are there and the values are the same.

xPaw commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
+	// ### `useHexIp`
 	//
-	leaveMessage: "The Lounge - https://thelounge.chat",
+	// When set to `true`, users' IP addresses will encoded has hex.
```
`has` -> `as`

And specify that this is encodes it in the `username` field and it only supports IPv4.

astorije replied:
Done, let me know what you think.

xPaw commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
+	// These three settings are used to run The Lounge using encrypted HTTP/2 on
+	// the server side. This will fallback to regular HTTPS if HTTP/2 is not
+	// supported.
 	//
```
Add a line saying reverse proxy is preferred if you want more options / live reload of certs

astorije replied:
Will do.

astorije replied:
Done, linked to the [upcoming guide on reverse proxies](https://5acafe274ed62f0a99ee7c04--thelounge-chat.netlify.com/docs/guides/reverse-proxies.html).

xPaw commented on scripts/generate-config-doc.js:
```diff
@@ -0,0 +1,67 @@
+
+	const generatedContent = `${infoBlock}\n\n${extractedDoc}\n${infoBlock}\n`;
+
+	fs.writeFileSync(docPath, generatedContent);
```
You wrap in async but use sync version here

astorije replied:
Yeah, I forget why but I used a promisified function above, and I couldn't be bothered to make everything beautifully async.
Async brings nothing here in terms of perf or whatnot, but I'll make the script consistent one way or the other.

astorije replied:
Done, all sync now. Clearer indeed. I'll finish addressing the other comment or 2 you left tomorrow.

xPaw commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
+	//   connections encrypted with TLS.
+	// - `password`
+	// - `tls`: Enable TLS connections
+	// - `rejectUnauthorized`: Whether or not the server certificate should be
```
`or not` is not necessary

xPaw commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
+	// - `password`
+	// - `tls`: Enable TLS connections
+	// - `rejectUnauthorized`: Whether or not the server certificate should be
+	//   verified against the list of supplied CAs by your Node.js installation.
```
Expand CAs to certificate authorities.

xPaw commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
-	//   3. Lounge tries to connect a second time, but this time using the user's
-	//      DN and password. Auth is validated iff this connection is successful.
+	// 1. The Lounge connects to the LDAP server with its system credentials.
+	// 2. It performs a LDAP search query to find the full DN associated to the
```
`an LDAP`?

astorije commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
 	//
-	// Hex IP
+	// This improves security and privacy by not exposing the client IP address,
+	// and always loading images from The Lounge and making all assets secure,
```
> This improves security and privacy by [...] always loading images from The Lounge and making all assets secure [...]

So `loading` is intended.

xPaw commented on defaults/config.js:
```diff
@@ -1,510 +1,428 @@
-	//
-	messageStorage: ["sqlite", "text"],
+	// This value is set to `["polling", "websocket"]` by default.
+	transports: ["polling", "websocket"],
```
It was a generic comment yeah, we shouldn't do this in this PR.

They mention making websocket default for socket.io 3.0 here https://github.com/socketio/socket.io/issues/3250

astorije replied:
Maybe open an issue to track this? (or maybe we already have one?)

astorije commented:
Alright @xPaw, I addressed all your comments. Let me know if there are other things you want changed.
@McInkay or @MaxLeiter, want to take another look?
